### PR TITLE
Trunk/mempool

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -6,12 +6,14 @@ on:
       - develop
       - master
       - releases/*
+      - trunk/mempool
```
Comment (Warchant) [inline .github/workflows/ci.yml]:
```suggestion
      - trunk/*
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -6,12 +6,14 @@ on:
       - develop
       - master
       - releases/*
```
Comment (Warchant) [inline .github/workflows/ci.yml]:
```suggestion
      - release/*
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -6,12 +6,14 @@ on:
...
+      - trunk/mempool
 
   pull_request:
     branches:
       - develop
       - master
       - releases/*
+      - trunk/mempool
```
Comment (Warchant) [inline .github/workflows/ci.yml]:
```suggestion
      - trunk/*
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -6,12 +6,14 @@ on:
...
 
   pull_request:
     branches:
       - develop
       - master
       - releases/*
```
Comment (Warchant) [inline .github/workflows/ci.yml]:
```suggestion
      - release/*
```

File: src/bootstraps.h [hunk 5]
```diff
@@ -41,6 +41,24 @@ struct AltChainParamsVBTC : public altintegration::AltChainParams {
...
+    {
+        return 1000000;
+    }
+
+    uint32_t getSuperMaxPopDataWeight() const noexcept override
+    {
+        return 5 * getMaxPopDataWeight();
+    }
+
+
```
Comment (Warchant) [inline src/bootstraps.h]: I think these are no longer relevant?

We have a hard limit of 1 MB per block, we should use that for limits

File: src/consensus/validation.h [hunk 6]
```diff
@@ -170,17 +170,15 @@ static inline int64_t GetTransactionWeight(const CTransaction& tx)
...
+
 static inline int64_t GetBlockWeight(const CBlock& block)
 {
-    int64_t txsPoPsize = 0;
-    for (const auto& tx : block.vtx)
-	{
-        if (VeriBlock::isPopTx(*tx)) {
-            txsPoPsize += GetTransactionWeight(*tx);
-        }
+    int64_t popDataSize = 0;
```
Comment (Warchant) [inline src/consensus/validation.h:176]: it is not size, it is weight

File: src/miner.cpp [hunk 7]
```diff
@@ -143,7 +142,13 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& sc
...
     int nDescendantsUpdated = 0;
 
-    addPackageTxs<VeriBlock::poptx_priority<ancestor_score>>(nPackagesSelected, nDescendantsUpdated, *pindexPrev);
+    addPackageTxs<ancestor_score>(nPackagesSelected, nDescendantsUpdated);
+
+    // VeriBlock: add PopData into the block
+    pblock->v_popData = VeriBlock::getService<VeriBlock::PopService>().getPopData(*pindexPrev);
+    if (!pblock->v_popData.empty()) {
+        pblock->nVersion |= VeriBlock::POP_BLOCK_VERSION_BIT;
+    }
```
Comment (Warchant) [inline src/miner.cpp:151]: fill pop data first, then fill regular txes

File: src/test/bloom_tests.cpp [hunk 8]
```diff
@@ -218,239 +218,239 @@ BOOST_AUTO_TEST_CASE(merkle_block_1)
...
+    //stream >> block;
```
Comment (Warchant) [inline src/test/bloom_tests.cpp]: uncomment this test

File: src/vbk/pop_service_impl.cpp [hunk 9]
```diff
@@ -149,75 +142,14 @@ PoPRewards PopServiceImpl::getPopRewards(const CBlockIndex& pindexPrev, const Co
         rewardValue >>= halvings;
         if ((rewardValue != 0) && (halvings < 64)) {
             CScript key = CScript(r.first.begin(), r.first.end());
-            btcRewards[key] = config.POP_REWARD_COEFFICIENT * rewardValue;
+            btcRewards[key] = rewardValue;
```
Comment (Warchant) [inline src/vbk/pop_service_impl.cpp]: Pls calculate POP_REWARD_COEFFICIENT as (100%-60%) * BASE_REWARD and multiply rewardValue by this coefficient.

File: src/vbk/pop_service_impl.cpp [hunk 10]
```diff
@@ -264,8 +191,7 @@ int PopServiceImpl::compareForks(const CBlockIndex& leftForkTip, const CBlockInd
     auto right = blockToAltBlock(rightForkTip);
     auto state = altintegration::ValidationState();
 
-    if (!altTree->setState(left.hash, state))
-    {
+    if (!altTree->setState(left.hash, state)) {
         return -1;
```
Comment (Warchant) [inline src/vbk/pop_service_impl.cpp:198]: should be 
```
if(!altTree->setState(left.hash, state)) {
  if(!altTree->setState(right.hash, state)) {
    throw std::logic_error("both chains are invalid");
  }
  
  return -1;
}
```

File: src/vbk/pop_service_impl.cpp [hunk 11]
```diff
@@ -276,221 +202,82 @@ PopServiceImpl::PopServiceImpl(const altintegration::Config& config)
...
-    altTree->removeSubtree(v);
+    return altTree->setState(block.asVector(), state);
 }
 
-bool PopServiceImpl::setState(const uint256& block, altintegration::ValidationState& state) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+std::vector<altintegration::PopData> PopServiceImpl::getPopData(const CBlockIndex& currentBlockIndex)
 {
-    AssertLockHeld(cs_main);
-    return altTree->setState(block.asVector(), state);
+    altintegration::AltBlock current = VeriBlock::blockToAltBlock(currentBlockIndex.nHeight, currentBlockIndex.GetBlockHeader());
```
Comment (Warchant) [inline src/vbk/pop_service_impl.cpp:220]: add AssertLockHeld(cs_main);

File: src/vbk/pop_service_impl.cpp [hunk 12]
```diff
@@ -276,221 +202,82 @@ PopServiceImpl::PopServiceImpl(const altintegration::Config& config)
...
-    auto& vbkp = *config.popconfig.vbk.params;
-    auto& btcp = *config.popconfig.btc.params;
-    altintegration::PopData popdata;
-
-    try {
-        while (pc < pend) {
-            if (!script.GetOp(pc, opcode, vchPushValue)) {
-                return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
-            }
+    mempool->removePayloads(v_popData);
```
Comment (Warchant) [inline src/vbk/pop_service_impl.cpp:228]: add AssertLockHeld(cs_main);

File: src/vbk/pop_service_impl.cpp [hunk 13]
```diff
@@ -276,221 +202,82 @@ PopServiceImpl::PopServiceImpl(const altintegration::Config& config)
...
-    // set return value of publications
-    if (pub != nullptr) {
-        pub->popData = popdata;
     }
 
     return true;
 }
 
-bool parseBlockPopPayloadsImpl(const CBlock& block, const CBlockIndex& pindexPrev, const Consensus::Params& params, BlockValidationState& state, std::vector<altintegration::AltPayloads>* payloads)
+bool popDataToPayloads(const CBlock& block, const CBlockIndex& indexPrev, BlockValidationState& state, std::vector<altintegration::AltPayloads>& payloads)
```
Comment (Warchant) [inline src/vbk/pop_service_impl.cpp:231]: once we add getHash(Endorsed block), this function no longer makes sense

File: src/bootstraps.h (hunk 5 above)
Comment (Mr-Leshiy) [reply src/bootstraps.h]: fixed

File: .github/workflows/ci.yml (hunk 4 above)
Comment (Mr-Leshiy) [reply .github/workflows/ci.yml]: fixed

File: .github/workflows/ci.yml (hunk 3 above)
Comment (Mr-Leshiy) [reply .github/workflows/ci.yml]: fixed

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Mr-Leshiy) [reply .github/workflows/ci.yml]: fixed

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Mr-Leshiy) [reply .github/workflows/ci.yml]: fixed

File: src/vbk/pop_service_impl.cpp (hunk 9 above)
Comment (Mr-Leshiy) [reply src/vbk/pop_service_impl.cpp]: fixed

File: src/vbk/pop_service_impl.cpp (hunk 10 above)
Comment (Mr-Leshiy) [reply src/vbk/pop_service_impl.cpp:198]: fixed

File: src/vbk/pop_service_impl.cpp (hunk 12 above)
Comment (Mr-Leshiy) [reply src/vbk/pop_service_impl.cpp:228]: fixed

File: src/vbk/pop_service_impl.cpp (hunk 11 above)
Comment (Mr-Leshiy) [reply src/vbk/pop_service_impl.cpp:220]: fixed

**Comment by Warchant** [comment]
pls push your changes

**Review by Warchant — APPROVED** [review]

**Pull request merged**